File: 2411-SpiralMatrixIv/2411-SpiralMatrixIv.js
// Last updated: 4/29/2026, 10:20:43 AM
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {number} m
 * @param {number} n
 * @param {ListNode} head
 * @return {number[][]}
 */
var spiralMatrix = function(m, n, head) {
    const mat = new Array(m);
    const visited = new Array(m);
    for(let i = 0; i < m; i++) {
        mat[i] = new Array(n).fill(-1);
        visited[i] = new Array(n).fill(false);
    }
    let dir = 0;
    const next = (x, y) => {
        return [[x, y+1], [x+1, y], [x, y-1], [x-1, y]];
    }
    const valid = (x, y) => {
        return x >=0 && x < m && y >= 0 && y < n;
    }

    const total = m * n;
    let x = 0, y = 0;
    let count = 0;
    let curr = head;
    mat[x][y] = curr.val;
    visited[x][y] = true;
    count++;
    curr = curr.next;
    while(curr && count < total) {
        const [x2, y2] = next(x, y)[dir];
        // console.log(`(x,y)=(${x},${y}), dir=${dir}, count=${count}, (x2, y2)=(${x2},${y2})`)
        if(valid(x2, y2) && !visited[x2][y2]) {
            // console.log(`valid & non-visited`)
            mat[x2][y2] = curr.val;
            visited[x2][y2] = true;
            curr = curr.next;
            count++;
            x = x2;
            y = y2;
        } else {
            // console.log(`change direction`)
            dir = (dir + 1) % 4;
        }
        // console.log(mat);
    }
    // console.log(mat);
    // console.log(visited);
    return mat;
};

File: 2411-SpiralMatrixIv/2411-SpiralMatrixIv.py
# Last updated: 4/29/2026, 10:20:48 AM
"""
11:23
O(m*n)
O(m*n)
"""
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def spiralMatrix(self, m: int, n: int, head: Optional[ListNode]) -> List[List[int]]:
        mat = [[-1]*n for i in range(m)]
        # print(mat)
        visited = [[False]*n for i in range(m)]
        def valid(i, j):
            return i >= 0 and i < m and j >= 0 and j < n and not visited[i][j]
        def next(i, j, direction):
            if direction == 0:
                return [i, j+1]
            elif direction == 1:
                return [i+1, j]
            elif direction == 2:
                return [i, j-1]
            else:
                return [i-1, j]
        
        node = head
        i, j = 0, 0
        direction = 0
        while node:
            mat[i][j] = node.val
            visited[i][j] = True
            next_i, next_j = next(i, j, direction)
            if valid(next_i, next_j):
                i, j = next_i, next_j
            else:
                direction = (direction + 1) % 4
                i, j = next(i, j, direction)
            node = node.next
        
        return mat